homeassistant/components/sms/__init__.py

```diff
@@ -17,17 +19,51 @@
 )
 
 
-def setup(hass, config):
+async def async_setup(hass, config):
     """Configure Gammu state machine."""
-    conf = config[DOMAIN]
-    device = conf.get(CONF_DEVICE)
-    gateway = gammu.StateMachine()  # pylint: disable=no-member
-    try:
-        gateway.SetConfig(0, dict(Device=device, Connection="at"))
-        gateway.Init()
-    except gammu.GSMError as exc:  # pylint: disable=no-member
-        _LOGGER.error("Failed to initialize, error %s", exc)
+    hass.data.setdefault(DOMAIN, {})
+    sms_config = config.get(DOMAIN, {})
+    if not sms_config:
+        return True
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=sms_config,
+        )
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Configure Gammu state machine."""
+
+    if SMS_GATEWAY in hass.data[DOMAIN]:
```

Comment:
Is there any suggested change with regards to `discovery.async_load_platform` ?
How/Where do you call this?

Reply:
Call it at the end of `async_setup_entry` in the base integration module.

Reply:
I am doing this:
```
async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
    """Configure Gammu state machine."""

    device = entry.data[CONF_DEVICE]
    config = {"Device": device, "Connection": "at"}
    gateway = await create_sms_gateway(config, hass)
    if not gateway:
        return False
    hass.data[DOMAIN][SMS_GATEWAY] = gateway
    hass.async_create_task(
        discovery.async_load_platform(hass, "notify", DOMAIN, {}, entry)
    )
    return True
```
But this fails because the Notify config is not compatible with the platform config.
What am I missing?

Reply:
The last argument should be the original whole config dict passed to `async_setup`. So store that in hass.data and use it when calling `async_load_platform`.

Reply:
I am trying, but I still get this:
```
2020-06-16 10:07:27 INFO (MainThread) [homeassistant.components.notify] Setting up notify.sms
2020-06-16 10:07:27 ERROR (MainThread) [homeassistant.components.notify] Error setting up platform sms
Traceback (most recent call last):
  File "/mnt/c/dd/Repos/home-assistant/homeassistant/components/notify/__init__.py", line 76, in async_setup_platform
    notify_service = await hass.async_add_job(
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/mnt/c/dd/Repos/home-assistant/homeassistant/components/sms/notify.py", line 26, in get_service
    number = config[CONF_RECIPIENT]
KeyError: 'recipient'
```

Reply:
The dict passed in async_setup is for the SMS platform, but the dict required for async_load_platform for "NOTIFY" is the dict for the notify.
I dont know how to get it from sms\__init__.py

Reply:
We need to update the `CONFIG_SCHEMA` in the base integration module. See eg:
https://github.com/home-assistant/core/blob/d278dd94777e602f23dce3690158fe414042b2fe/homeassistant/components/netgear_lte/__init__.py#L92-L94

https://github.com/home-assistant/core/blob/d278dd94777e602f23dce3690158fe414042b2fe/homeassistant/components/netgear_lte/__init__.py#L60-L65

Note that this will be a breaking change.

Reply:
> The dict passed in async_setup is for the SMS platform, but the dict required for async_load_platform for "NOTIFY" is the dict for the notify.
> I dont know how to get it from sms__init__.py
> 
> Do I need to change get_service in notify.py ?
> 
> ```
> def get_service(hass, config, discovery_info=None):
>     """Get the SMS notification service."""
>     if SMS_GATEWAY not in hass.data[DOMAIN]:
>         _LOGGER.error("SMS gateway not found, cannot initialize service")
>         return
>     gateway = hass.data[DOMAIN][SMS_GATEWAY]
>     number = config[CONF_RECIPIENT]
>     return SMSNotificationService(gateway, number)
> ```

I figure it out. Please review newest change. Now when unloading and reloading, the notify service will be reloaded and no need to restart hass. :)

Reply:
> We need to update the `CONFIG_SCHEMA` in the base integration module. See eg:
> https://github.com/home-assistant/core/blob/d278dd94777e602f23dce3690158fe414042b2fe/homeassistant/components/netgear_lte/__init__.py#L92-L94
> 
> https://github.com/home-assistant/core/blob/d278dd94777e602f23dce3690158fe414042b2fe/homeassistant/components/netgear_lte/__init__.py#L60-L65
> 
> Note that this will be a breaking change.
> 
> Then pass the required info as discovery_info or alternatively store it in hass.data:
> https://github.com/home-assistant/core/blob/d278dd94777e602f23dce3690158fe414042b2fe/homeassistant/components/netgear_lte/__init__.py#L227-L239

I did it without having to introduce a breaking change. Is this way to do it ok?

Reply:
The point with loading via discovery is that we can control that there's a config entry ready before setting up the platform. If we keep the platform config set up path there's no point with the discovery path.

Reply:
So what you are saying is that the config will need to change from this:
```
notify:
  - name: sms_person1
    platform: sms
    recipient: "+15551234"
  - name: sms_person2
    platform: sms
    recipient: "+15554321"
```

To this:
```
sms:
  notify:
    - name: sms_person1
      recipient: "+15551234"
    - name: sms_person2
      recipient: "+15554321"
```

Does this mean we cannot keep both kinds of configs?
Do we really need to force the breaking change?

Reply:
If we can start the config flow from the notify platform if the config entry is missing, I think option 3 is viable.